cpp/src/replace/replace.cu

```diff
@@ -0,0 +1,209 @@
+ * limitations under the License.
+ */
+#include <thrust/device_ptr.h>
+#include <thrust/find.h>
+#include <thrust/execution_policy.h>
+#include <cub/cub.cuh>
+
+#include <cudf/legacy/copying.hpp>
+#include <cudf/legacy/replace.hpp>
+#include <cudf/cudf.h>
+#include <rmm/rmm.h>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/legacy/type_dispatcher.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <utilities/cudf_utils.h>
+#include <utilities/cuda_utils.hpp>
+#include <utilities/column_utils.hpp>
+#include <bitmask/legacy/legacy_bitmask.hpp>
+#include <bitmask/legacy/bit_mask.cuh>
+
+namespace {  // anonymous
+
+using namespace cudf;
+
+static constexpr int BLOCK_SIZE = 256;
+
+/* --------------------------------------------------------------------------*/
+/**
+ * @brief Kernel that converts inputs from `in` to `out`  using the following
+ *        rule:   Convert  -NaN  -> NaN
+ *                Convert  -0.0  -> 0.0
+ *
+ * @param[in] column_device_view representing input data
+ * @param[in] mutable_column_device_view representing output data. can be
+ *            the same actual underlying buffer that in points to. 
+ *
+ * @returns
+ */
+/* ----------------------------------------------------------------------------*/
+template <typename T>
+__global__
+void normalize_nans_and_zeros(column_device_view in, 
+                              mutable_column_device_view out)
+{
+   int tid = threadIdx.x;
+   int blkid = blockIdx.x;
+   int blksz = blockDim.x;
+   int gridsz = gridDim.x;
+
+   int start = tid + blkid * blksz;
+   int step = blksz * gridsz;
+
+   // grid-stride
+   for (int i=start; i<in.size(); i+=step) {
+      if(!in.is_valid(i)){
+         continue;
+      }
+
+      T el = in.element<T>(i);
+      if(std::isnan(el)){
+         out.element<T>(i) = std::numeric_limits<T>::quiet_NaN();
+      } else if(el == (T)-0.0){
+         out.element<T>(i) = (T)0.0;
+      } else {
+         out.element<T>(i) = el;
+      }
+   }
+}                        
+
+  /* --------------------------------------------------------------------------*/
+  /**
+   * @brief Functor called by the `type_dispatcher` in order to invoke and instantiate
+   *        `normalize_nans_and_zeros` with the appropriate data types.
+   */
+  /* ----------------------------------------------------------------------------*/
+struct normalize_nans_and_zeros_kernel_forwarder {
+   // floats and doubles. what we really care about.
+   template <typename T, std::enable_if_t<std::is_floating_point<T>::value>* = nullptr>
+   void operator()(  column_device_view in,
+                     mutable_column_device_view out,
+                     cudaStream_t stream)
+   {
+      util::cuda::grid_config_1d grid{in.size(), BLOCK_SIZE};
+      normalize_nans_and_zeros<T><<<grid.num_blocks, BLOCK_SIZE, 0, stream>>>(in, out);
```

Comment:
Couple of things to simplify this.

First, let's use `thrust::transform/transform_if` instead of a custom kernel.

Second, let's specialize code paths for whether or not null value exist.

End result looks something like this:
```
struct normalize_nans_and_zeros_kernel_forwarder {
  // floats and doubles. what we really care about.
  template <typename T,
            std::enable_if_t<std::is_floating_point<T>::value>* = nullptr>
  void operator()(column_device_view in, mutable_column_device_view out,
                  cudaStream_t stream) {
    auto normalize_nans_zeros = [in] __device__(auto i) {
      auto e = in.element<T>();
      if (isnan(e)) {
        return std::numeric_limits<T>::quiet_NaN();
      }
      if (T{0.0} == e) {
        return T{0.0};
      }
    };

    auto is_valid = [in] __device__(auto i) { return in.is_valid_nocheck(i); }

    if (in.has_nulls()) {
      thrust::transform_if(rmm::exec_policy(stream)->on(stream),
                           thrust::make_counting_iterator(0),
                           thrust::make_counting_iterator(in.size()),
                           out.begin<T>(), normalize_nans_zeros, is_valid);
    }
    else {
      thrust::transform(rmm::exec_policy(stream)->on(stream),
                        thrust::make_counting_iterator(0),
                        thrust::make_counting_iterator(in.size()),
                        out.begin<T>());
    }
  }
};
```

Reply:
Actually, there's no need to check for nulls. It should be fine to overwrite whatever garbage is in memory for a null value. 

So it's even easier:
```
struct normalize_nans_and_zeros_kernel_forwarder {
  // floats and doubles. what we really care about.
  template <typename T,
            std::enable_if_t<std::is_floating_point<T>::value>* = nullptr>
  void operator()(column_device_view in, mutable_column_device_view out,
                  cudaStream_t stream) {
    auto normalize_nans_zeros = [in] __device__(auto i) {
      auto e = in.element<T>();
      if (isnan(e)) {
        return std::numeric_limits<T>::quiet_NaN();
      }
      if (T{0.0} == e) {
        return T{0.0};
      }
    };
      thrust::transform(rmm::exec_policy(stream)->on(stream),
                        thrust::make_counting_iterator(0),
                        thrust::make_counting_iterator(in.size()),
                        out.begin<T>(), normalize_nans_zeros);
    
  }
};
```

Reply:
Fixed in https://github.com/rapidsai/cudf/pull/3291/commits/1c3389e41ce7e87c5bf35a6befe154b55ef8d220